src/threads/ThreadState.cpp

```diff
@@ -53,45 +146,114 @@ int Level::getMaxThreadsAtNextLevel() const
+            nowaitCv->wait(lock);
         }
 
         // Reset, after we've finished
-        nowaitCount = 0;
+        nowaitCount->store(0);
     } else {
-        // Notify master that this thread has finished and continue
-        nowaitCount++;
-        nowaitCv.notify_one();
+        // If this is the last non-master thread, notify
+        int countBefore = nowaitCount->fetch_add(1);
+        if (countBefore == numThreads - 2) {
+            nowaitCv->notify_one();
+        }
+    }
+}
+
+std::vector<uint8_t> Level::serialise()
+{
+    // Work out the size of this object
+    size_t thisSize = sizeof(Level);
+    thisSize += nSharedVars * sizeof(uint32_t);
+
+    uint8_t* bytesPtr = BYTES(this);
+    std::vector<uint8_t> bytes(bytesPtr, bytesPtr + thisSize);
+    return bytes;
+}
+
+void Level::deserialise(const Level* other)
+{
+    id = other->id;
+    depth = other->depth;
+    activeLevels = other->activeLevels;
+    maxActiveLevels = other->maxActiveLevels;
+    numThreads = other->numThreads;
+    pushedThreads = other->pushedThreads;
+    wantedThreads = other->wantedThreads;
+
+    globalTidOffset = other->globalTidOffset;
+
+    nSharedVars = other->nSharedVars;
+    sharedVars = new uint32_t[nSharedVars];
+    std::memcpy(sharedVars, other->sharedVars, nSharedVars * sizeof(uint32_t));
+}
+
+void Level::waitOnBarrier()
+{
+    if (numThreads > 1) {
+        FROM_MAP(b, faabric::util::Barrier, barriers, numThreads);
+        b->wait();
     }
 }
 
-void setUpOpenMPContext(const int threadNum, std::shared_ptr<Level>& thisLevel)
+void Level::lockCritical()
+{
+    FROM_MAP(mx, std::recursive_mutex, levelMutexes);
+    mx->lock();
+}
+
+void Level::unlockCritical()
 {
-    OpenMPContext& ctx = getOpenMPContext();
-    ctx.threadNumber = threadNum;
-    ctx.level = thisLevel;
+    FROM_MAP(mx, std::recursive_mutex, levelMutexes);
+    mx->unlock();
 }
 
-void setUpOpenMPContext(const faabric::Message& msg)
+// Note that we need be able to translate between local and global thread
+// numbers. The global thread number must be unique in the system, while the
+// local thread number must fit with that expected by OpenMP within the team/
+// level. We use Faabric messages to hold the global thread number and can
+// translated back and forth.
+int Level::getLocalThreadNum(faabric::Message* msg)
 {
-    OpenMPContext& ctx = getOpenMPContext();
-    ctx.threadNumber = msg.ompthreadnum();
+    if (depth == 0) {
+        return msg->appindex();
+    }
+
+    int localThreadNum = msg->appindex() - globalTidOffset;
 
-    // Floor max active levels at 1 for a message. Max active levels of zero
-    // disables parallel code.
-    int maxActiveLevels = std::max(msg.ompmal(), 1);
+    if (localThreadNum < 0) {
+        faabric::util::getLogger()->error(
+          "Local thread num negative: {} - {} @ {}",
+          msg->appindex(),
+          globalTidOffset,
+          depth);
+    }
+
+    assert(localThreadNum >= 0);
```

Comment:
Checking, throwing an error, and asserting seems a bit counter-intuitive/verbose to me. What do you think about using something like:
```cpp
#define assertm(exp, msg) assert(((void)msg, exp))
...
assertm(localThreadNum >= 0, fmt::format("Local thread num negative {} - {} @ {}", msg->appindex(), globalTidOffset, depth));
```
taken from [cppreference](https://en.cppreference.com/w/cpp/error/assert).

Reply:
Hmm yes you're right, I'll just add an exception in this case.